src/main/java/com/palantir/ri/ResourceIdentifier.java

```diff
@@ -315,9 +348,115 @@ private static void checkTypeIsValid(String type) {
         }
     }
 
-    private static void checkLocatorIsValid(String value) {
-        if (!isValidLocator(value)) {
-            throw new IllegalArgumentException("Illegal locator format: " + value);
+    private static void checkLocatorIsValid(String locator) {
+        if (!isValidLocator(locator)) {
+            throw new IllegalArgumentException("Illegal locator format: " + locator);
+        }
+    }
+
+    private static int getServiceIndex(String value, int start) {
+        if (value == null) {
+            return Index.INVALID;
         }
+
+        if (start + 1 > value.length()) {
+            return Index.INVALID;
+        }
+
+        for (int i = start; i < value.length(); i++) {
+            char ch = value.charAt(i);
+            if (i == start) {
+                if (!isLowerAlpha(ch)) {
+                    return Index.INVALID;
+                }
+            } else if (ch == SEPARATOR) {
+                return i;
+            } else if (!(isLowerAlpha(ch) || isDigit(ch) || isDash(ch))) {
+                return Index.INVALID;
+            }
+        }
+
+        return Index.END;
+    }
+
+    @SuppressWarnings("CyclomaticComplexity")
+    private static int getInstanceIndex(String value, int start) {
+        if (value == null) {
+            return Index.INVALID;
+        }
+
+        if (start > value.length()) {
```

Comment:
Is there any value to hoisting the length out to a variable so we don't call `.length()` on every loop? Or should we assume that since it almost certainly will be inlined that it's just a memory access either way?

Reply:
At minimum it would reduce the bytecode size so this is more likely to be inlined. Reading object fields, even final ones, is often more expensive than one might expect. Best if we read once for both the loop and conditional above.